scylla-cdc-replicator/src/replicator_consumer.rs

```diff
@@ -0,0 +1,237 @@
+use anyhow::anyhow;
+use async_trait::async_trait;
+use itertools::Itertools;
+use scylla::frame::response::result::CqlValue;
+use scylla::prepared_statement::PreparedStatement;
+use scylla::query::Query;
+use scylla::transport::topology::{ColumnKind, CqlType, Table};
+use scylla::Session;
+use scylla_cdc::consumer::*;
+use std::slice::Iter;
+use std::sync::Arc;
+
+struct ReplicatorConsumer {
+    dest_session: Arc<Session>,
+    dest_keyspace_name: String,
+    dest_table_name: String,
+    // [TODO]: Not sure, if we want to keep the ownership here,
+    //   or just a ref and keep it in the factory.
+    table_schema: Table,
+    other_columns: Vec<String>,
+
+    // Prepared queries.
+    insert_query: PreparedStatement,
+}
+
+impl ReplicatorConsumer {
+    async fn new(
+        dest_session: Arc<Session>,
+        dest_keyspace_name: String,
+        dest_table_name: String,
+        table_schema: Table,
+    ) -> ReplicatorConsumer {
+        // [TODO]: This might be extracted to Factory or in some other place,
+        //      in order to not generate it with every consumer created.
+
+        // Iterator for both: partition keys and clustering keys.
+        let keys_iter = ReplicatorConsumer::get_keys_iter(&table_schema).await;
+        // Collect names of columns that are not clustering or partition key.
+        let other_columns = table_schema
+            .columns
+            .iter()
+            .filter(|column| {
+                column.1.kind != ColumnKind::Clustering && column.1.kind != ColumnKind::PartitionKey
+            })
+            .map(|column| column.0.clone())
+            .collect::<Vec<String>>();
+
+        // Clone, because the iterator is consumed.
+        let names = keys_iter.clone().join(",");
+        let markers = keys_iter
+            .map(|col_name| {
+                ReplicatorConsumer::bind_marker_for_type(
+                    &table_schema
+                        .columns
+                        .get(col_name)
+                        .expect("Column name not found")
+                        .type_,
+                )
+            })
+            .join(",");
+
+        let insert_query = dest_session
+            .prepare(format!(
+                "INSERT INTO {}.{} ({}) VALUES ({}) USING TTL ?",
+                dest_keyspace_name, dest_table_name, names, markers
+            ))
+            .await
+            .expect("Preparing insert query failed.");
+
+        ReplicatorConsumer {
+            dest_session,
+            dest_keyspace_name,
+            dest_table_name,
+            table_schema,
+            other_columns,
+            insert_query,
+        }
+    }
+
+    async fn get_keys_iter(
+        table_schema: &Table,
+    ) -> std::iter::Chain<Iter<'_, String>, Iter<'_, String>> {
```

Comment:
Tip: you can simplify the return type using [`impl Trait`](https://doc.rust-lang.org/reference/types/impl-trait.html):

```rust
) -> impl std::iter::Iterator<Item=String> {
```

Reply:
I tried to do so, but I think it should iterate over `&String`, not `String`. Nevertheless, the clones don't work with this solution:
```
error[E0599]: no method named `clone` found for opaque type `impl Iterator` in the current scope
  --> scylla-cdc-replicator/src/replicator_consumer.rs:49:31
   |
49 |         let names = keys_iter.clone().join(",");
   |                               ^^^^^ method not found in `impl Iterator`
```

Reply:
How about `impl std::iter::Iterator<Item=&String> + Clone`, then?

Reply:
It compiles, I've just pushed this (see below).